pkg/nsx/services/securitypolicy/firewall.go

```diff
@@ -706,26 +710,64 @@ func (service *SecurityPolicyService) deleteSecurityPolicy(sp types.UID) error {
 		return err
 	}
 
-	err = securityPolicyStore.Apply(&finalSecurityPolicyCopy)
+	// Apply security policy and rules to store
+	err = service.applySecurityPolicyStore(finalSecurityPolicyCopy, finalSecurityPolicyCopy.Rules, *nsxGroupsEmpty, true)
 	if err != nil {
-		log.Error(err, "Failed to apply store", "securityPolicy", finalSecurityPolicyCopy)
 		return err
 	}
-	err = ruleStore.Apply(&finalSecurityPolicyCopy.Rules)
-	if err != nil {
-		log.Error(err, "Failed to apply store", "nsxRules", finalSecurityPolicyCopy.Rules)
-		return err
-	}
-	err = groupStore.Apply(nsxGroups)
+
+	// Delete groups after security policy and rules are deleted
+	err = service.deleteGroupsWithStore(nsxGroups, nil, groupStore)
 	if err != nil {
-		log.Error(err, "Failed to apply store", "nsxGroups", nsxGroups)
 		return err
 	}
 
 	log.Info("Successfully deleted NSX SecurityPolicy", "nsxSecurityPolicy", finalSecurityPolicyCopy)
 	return nil
 }
 
+// deleteGroupsWithStore deletes NSX groups from NSX and the group store.
+// It handles ResourceInUse errors by logging them but continuing execution.
+// Returns an error if deletion fails with a non-ResourceInUse error or if applying to the store fails.
+func (service *SecurityPolicyService) deleteGroupsWithStore(nsxGroups *[]model.Group, vpcInfo *common.VPCResourceInfo, groupStore *GroupStore) error {
+	if len(*nsxGroups) == 0 {
+		return nil
+	}
+
+	for _, group := range *nsxGroups {
+		groupID := *group.Id
+		var err error
+
+		// Use the appropriate client based on whether VPC is enabled
+		if IsVPCEnabled(service) {
```

Comment:
You can mix group delete for VPC mode and T1 mode together. 
Also, please separate nsx client deletion logic from group store apply. We don't want both mixed in one function.

Reply:
[split DeleteGroupsWithStore for T1 and VPC](https://github.com/vmware-tanzu/nsx-operator/pull/1159/commits/215d5a8d94256cc6cff1cc21e866b254a2273104)